### PR TITLE
Fixes #7913: list-compatible input should be able to filter on minimum and maximum version

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/7913

### CODE REVIEW

File: tree/10_ncf_internals/list-compatible-inputs [hunk 1]
```diff
@@ -58,18 +58,29 @@ do
...
+      if grep -q "${tag_regex}<" "${file}"; then operator="<"; fi
+      major=`sed -ne "s/${tag_regex}${operator}${version_regex}/\\1/p" "${file}"`
+      minor=`sed -ne "s/${tag_regex}${operator}${version_regex}/\\2/p" "${file}"`
       file_name=`basename ${file}`
-      if [ "${major}" -gt "${cfengine_major}" ] || [ "${major}" -eq "${cfengine_major}" -a "${minor}" -gt "${cfengine_minor}" ]
+      if [ "${operator}" = ">=" ]
       then
-        # posix compliant syntax to exclude a file
-        excludes="${excludes} -name ${file_name} -prune -o"
+        # exclude if minimum_version>cfengine_version
```
Comment (jooooooon) [inline tree/10_ncf_internals/list-compatible-inputs]: Above you grep for the operator ```>=```, but this comment says ```>``` (this is a minor detail but does hinder understanding)

File: tree/10_ncf_internals/list-compatible-inputs [hunk 2]
```diff
@@ -58,18 +58,29 @@ do
...
+        if [ "${major}" -gt "${cfengine_major}" ] || [ "${major}" -eq "${cfengine_major}" -a "${minor}" -gt "${cfengine_minor}" ]
+        then
+          # posix compliant syntax to exclude a file
+          excludes="${excludes} -name ${file_name} -prune -o"
+        fi
+      elif [ "${operator}" = "<" ]
+        # exclude if minimum_version<=cfengine_version
+        if [ "${major}" -lt "${cfengine_major}" ] || [ "${major}" -eq "${cfengine_major}" -a "${minor}" -le "${cfengine_minor}" ]
```
Comment (jooooooon) [inline tree/10_ncf_internals/list-compatible-inputs]: In this case, the comment also differs from the grep above (```<=``` vs ```<```)
(but the implementation seems to match ```<```)

File: tree/10_ncf_internals/list-compatible-inputs [hunk 3]
```diff
@@ -58,18 +58,29 @@ do
...
-    tag_regex='^#[ \t]*@agent_version[ \t]*>='
+    tag_regex='^#[ \t]*@agent_version[ \t]*'
     for file in `find "${framework_path}/${directory}" -name '*.cf' -exec grep -l "${tag_regex}" '{}' \;`
     do
-      major=`sed -ne "s/${tag_regex}${version_regex}/\\1/p" "${file}"`
-      minor=`sed -ne "s/${tag_regex}${version_regex}/\\2/p" "${file}"`
+      if grep -q "${tag_regex}>=" "${file}"; then operator=">="; fi
+      if grep -q "${tag_regex}<" "${file}"; then operator="<"; fi
+      major=`sed -ne "s/${tag_regex}${operator}${version_regex}/\\1/p" "${file}"`
+      minor=`sed -ne "s/${tag_regex}${operator}${version_regex}/\\2/p" "${file}"`
```
Comment (jooooooon) [inline tree/10_ncf_internals/list-compatible-inputs]: This is not strictly related to this PR, but I think it would be useful to allow for whitespace like this:

    major=`sed -ne "s/${tag_regex}${operator}[ \t]*${version_regex}/\\1/p" "${file}"`

File: tree/10_ncf_internals/list-compatible-inputs (hunk 1 above)
Comment (peckpeck) [inline tree/10_ncf_internals/list-compatible-inputs]: Yes, the thing is, this matches the behaviour on the line below.
It has been chosen because it is a negative match, and I didn't want to change the line

File: tree/10_ncf_internals/list-compatible-inputs (hunk 2 above)
Comment (peckpeck) [inline tree/10_ncf_internals/list-compatible-inputs]: The comment is wrong, indeed

**Comment by peckpeck** [comment]
PR updated

File: tree/10_ncf_internals/list-compatible-inputs (hunk 1 above)
Comment (jooooooon) [inline tree/10_ncf_internals/list-compatible-inputs]: Wow, indeed, you're right. This is clearly a bug. I suggest you open a ticket for it and correct it separately (I agree, this PR should not change that)

Comment (peckpeck) [inline tree/10_ncf_internals/list-compatible-inputs]: It is not a bug, just a choice in how to write the bash code.

File: tree/10_ncf_internals/list-compatible-inputs [hunk 4]
```diff
@@ -58,18 +58,29 @@ do
...
+      if grep -q "${tag_regex}<" "${file}"; then operator="<"; fi
+      major=`sed -ne "s/${tag_regex}${operator}[ \\t]*${version_regex}/\\1/p" "${file}"`
+      minor=`sed -ne "s/${tag_regex}${operator}[ \\t]*${version_regex}/\\2/p" "${file}"`
       file_name=`basename ${file}`
-      if [ "${major}" -gt "${cfengine_major}" ] || [ "${major}" -eq "${cfengine_major}" -a "${minor}" -gt "${cfengine_minor}" ]
+      if [ "${operator}" = ">=" ]
       then
-        # posix compliant syntax to exclude a file
-        excludes="${excludes} -name ${file_name} -prune -o"
+        # exclude if requirered_version<cfengine_version
```
Comment (jooooooon) [inline tree/10_ncf_internals/list-compatible-inputs]: I think you mean if cfengine_version < required_version

(also, required not requirered :) )

Comment (jooooooon) [inline tree/10_ncf_internals/list-compatible-inputs]: Damn this is ******** confusing!!!

I hate code like this, we're simply comparing a version number and it's taking me 30 minutes to understand the code. I will propose a separate PR to rewrite this in a way that is a lot easier to understand.

Comment (jooooooon) [inline tree/10_ncf_internals/list-compatible-inputs]: But yes, you're right of course, this is not a bug.

**Comment by peckpeck** [comment]
PR updated
